Sources/AsyncHTTPClient/ConnectionPool/HTTP2/HTTP2ClientRequestHandler.swift

```diff
@@ -239,21 +239,29 @@ final class HTTP2ClientRequestHandler: ChannelDuplexHandler {
 
     private func runSuccessfulFinalAction(_ action: HTTPRequestStateMachine.Action.FinalSuccessfulRequestAction, context: ChannelHandlerContext) {
         switch action {
-        case .close:
+        case .close, .none:
+            // We must close the http2 stream after the request has finished. This breaks a reference
+            // cycle in HTTP2Connection.
             context.close(promise: nil)
 
         case .sendRequestEnd(let writePromise):
-            context.writeAndFlush(self.wrapOutboundOut(.end(nil)), promise: writePromise)
-
-        case .none:
-            break
+            let promise = writePromise ?? context.eventLoop.makePromise(of: Void.self)
+            context.writeAndFlush(self.wrapOutboundOut(.end(nil)), promise: promise)
+            promise.futureResult.whenComplete { _ in
+                // Once we have written the last message we must close the http2 stream, since this
+                // break a reference cycle in HTTP2Connection.
+                context.close(promise: nil)
```

Comment:
I'm a little surprised that this is necessary. swift-nio-http2 should notify you that the stream is closed when both sides have sent their `.end`. Are you entirely confident this is required?

Reply:
I'm not. I was only testing the child channel. And wanted to be as defensive as possible. Do you want me to remove it again?

Reply:
Yeah, I think we shouldn't add code that we don't really need.